posix: sysconf: match _SC* defines with newlib/picolib

nslowell opened the pull request:
It's possible for newlib/picolib libc libraries to internally call sysconf() which would execute zephyr's implementation.  However, if the _SC* defines do not have matching values, then the incorrect switch case executes.

This issue arises when using newlib/picolib libc that includes sysconf implementation for ARM.  With current defaults, the zephyr sysconf() overrides the original libc sysconf() so we must ensure proper operation.

keith-packard commented:
Is there some reason not to just use the picolibc/newlib values when those C libraries are in use?

cfriedt reviewed (commented):
@nslowell - can we switch to `#define`s instead of `enum`?

The major benefit of that is getting warnings / errors if the symbol is redefined.

But also, the enum was mainly there because originally there was no specific ordering required.

keith-packard commented:
> @nslowell - can we switch to `#define`s instead of `enum`?

And use the #defines that come from libc when available?

cfriedt commented on include/zephyr/posix/sys/sysconf.h:
```diff
@@ -280,7 +306,6 @@ enum {
 #define __z_posix_sysconf_SC_MQ_PRIO_MAX                  MQ_PRIO_MAX
 #define __z_posix_sysconf_SC_OPEN_MAX                     CONFIG_ZVFS_OPEN_MAX
 #define __z_posix_sysconf_SC_PAGE_SIZE                    PAGE_SIZE
-#define __z_posix_sysconf_SC_PAGESIZE                     PAGESIZE
```
This might still need to be present

cfriedt commented:
> And use the #defines that come from libc when available?

That could be achieved with something like `#include_next <sys/sysconf.h>` assuming that file exists in newlib / picolibc.

keith-packard commented:
> That could be achieved with something like `#include_next <sys/sysconf.h>` assuming that file exists in newlib / picolibc.

Yup, although the standard file is <unistd.h>, so Zephyr should use that if possible. If that's not possible, then using <sys/unistd.h> could work.

nslowell commented:
Appreciate the feedback @cfriedt and @keith-packard.  

Switching from enum to #defines that match newlib/picolib seemed doable for POSIX_SYSCONF_IMPL_FULL. I did have to adjust how z_sysconf() is used to prevent it from breaking due to preprocessor expansion, like I mentioned in https://github.com/zephyrproject-rtos/zephyr/discussions/87833#discussioncomment-12690762
Example:
```
#define z_sysconf(x) (long)CONCAT(__z_posix_sysconf_, x) /* <--- add trailing underscore here */

switch(x) {
	case _SC_ADVISORY_INFO:
		return z_sysconf(SC_ADVISORY_INFO); /* <--- drop the leading underscore here */
```

However, for POSIX_SYSCONF_IMPL_MACRO, there is no way around the sysconf breakage, so the #defines have to be wrapped within #ifdef POSIX_SYSCONF_IMPL_FULL.

Secondly, I really thought getting this code to reference newlib/picolib's unistd.h was going to be a bit of a headache, but I guess I got lucky.  It's possible there's some scenario this breaks, but it seemed to work with all libc options (minimal, newlib, and picolib).  

I'm hoping https://github.com/zephyrproject-rtos/zephyr/pull/88195 will merge soon, so this can run the full test suite and find any potential issues with this change.

Thanks again for the feedback.

Commit:
posix: net: move static inline gethostname() to lib definition

static inline gethostname() in unistd.h can cause declaration collisions.
we should just move it to a normal function definition like the
rest of the network functions.

Signed-off-by: Nicholas Lowell <Nicholas.Lowell@lexmark.com>

Commit:
posix: move PTHREAD_STACK_MIN to posix_features

PTHREAD_STACK_MIN is one of the very few posix features
referenced in sysconf that is not defined in posix_features.h
Move it to minimize header dependencies.

Signed-off-by: Nicholas Lowell <Nicholas.Lowell@lexmark.com>

Commit:
posix: device_io: mark pwrite() buf as const

match standard by having input void *buf parameter
for pwrite() marked as const, and avoid any potential
declaration conflicts

Signed-off-by: Nicholas Lowell <Nicholas.Lowell@lexmark.com>

Commit:
posix: sysconf: match _SC* defines with newlib/picolib

It's possible for newlib/picolib libc libraries to
internally call sysconf() which would execute zephyr's
implementation.  However, if the _SC* defines do not have
matching values, then the incorrect switch case executes.

This issue arises when using newlib/picolib libc that includes
sysconf implementation for ARM.  With current defaults, the
zephyr sysconf() overrides the original libc sysconf() so
we must ensure proper operation.

We will switch to the #define list just like newlib/picolib.
We can't currently use their unistd.h directly due to a domino
of declaration conflicts.
For the "small" macro implementation, we have to drop using
CONCAT to prevent pre-expansion of the new #defines

Signed-off-by: Nicholas Lowell <Nicholas.Lowell@lexmark.com>

nslowell commented:
@keith-packard, @cfriedt 
As I had feared, trying to #include_next <unistd.h> to pull in defines from newlib/picolib creates a mountain of declaration conflicts I simply was not able to climb over.  This PR now has a few commits, the first few being an initial attempt to resolve issues but they just continued to pile up.

Secondly, there's no way to maintain the macro version of sysconf() and have successful tests without isolating it and keeping the original enum declaration for it.  The new #defines that match newlib/picolib values will be only for the full implementation.

Hope this is sufficient.  I've got all tests passing (except for timeouts which are unrelated and pass when run locally)

cfriedt commented on include/zephyr/posix/sys/sysconf.h:
```diff
@@ -296,6 +331,134 @@ enum {
 #define __z_posix_sysconf_SC_TZNAME_MAX                   TZNAME_MAX
 
 #ifdef CONFIG_POSIX_SYSCONF_IMPL_MACRO
+enum {
```
Hmm... having these defined two ways is kind of sketchy.

cfriedt replied:
@keith-packard - can you think of any macrobatics that would simplify `CONFIG_POSIX_SYSCONF_IMPL_MACRO`? Some kind of non-evaluative token pasting?

Otherwise, I'm thinking it might be best to go back to the individually numbered enums. Sorry to ask @nslowell to do more work :(

nslowell replied:
I asked myself this several times, but I'm glad you asked again on a new day b/c I discovered that #define sysconf() exists in unistd.h and sysconf.h and I was testing changes on the wrong one.  Preliminary tests show me that we actually want to _avoid_ any double-layer macro, so
`#define sysconf(x) __z_posix_sysconf##x` 
actually seems to work.  I'll make the official change and run through tests again tomorrow.

I appreciate the continued support and patience.

cfriedt replied:
Ah, ok - that's how it is now, but we use the `CONCAT()` macro

nslowell replied:
right, we can't use CONCAT() b/c it causes the _SC_* to be pre-expanded.  By concatenting directly in the sysconf() macro, the pre-expansion is avoided.

nslowell commented:
@cfriedt @keith-packard  I think this is good to go now.

cfriedt commented on include/zephyr/posix/unistd.h:
```diff
@@ -72,7 +64,8 @@ size_t confstr(int name, char *buf, size_t len);
 #ifdef CONFIG_POSIX_SYSCONF_IMPL_MACRO
-#define sysconf(x) (long)CONCAT(__z_posix_sysconf, x)
+/* Can't use CONCAT(), must concat directly to prevent expansion of 'x' */
+#define sysconf(x) (long)__z_posix_sysconf##x
```
In case there is another revision, I would just bracket `(...)` this entire expression out of paranoia, but not blocking.

cfriedt reviewed (approved):
LGTM - thanks for all of your help @nslowell 🙏
